CARDS-2286: Allow to configure token life time for each survey

veronikaslc opened the pull request:
Testing "daysRelativeToEventWhileSurveyIsValid": "Relatively to the associated event, patients can fill out surveys within":
* Start in prems
* create a clinic A with a valid `survey` and leave  the `daysRelativeToEventWhileSurveyIsValid` field blank
* create a clinic B with a valid `survey` and set the `daysRelativeToEventWhileSurveyIsValid` field to 2
* create 2 patient visits for 2 different patients, one at clinic A and one at clinic B
  * check the autocreated Survey events forms, make sure the link expiry date field reflects the token lifetime value (for the clinic A visit, it should be in 29 days (if visit date is yesterday), for the clinic B visit it should be tomorrow (if the visit date is yesterday)
  * generate tokens for both visits and go to the respective survey links to check the link expiration message displayed on the start screen - it should say that the link will expire in 29 days for clinic A and in 1 day for clinic B
* From /bin/browser, edit the daysRelativeToEventWhileSurveyIsValid of clinic B to be 0
  * generate another token for the visit to clinic B and go to the survey link - it should say the link has expired
* testing emailed links - TBD
* Unsubmitted forms cleanup
  * reconfigure the unsubmitted forms cleanup task to run every 2min
  * test that it deletes the forms for clinic B, but NOT for clinic A

Commit:
CARDS-2286: Allow to configure token life time for each survey

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Move tokenLifetime to the clinic

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/QuestionnaireSet.jsx:
```diff
@@ -582,7 +584,7 @@ function QuestionnaireSet(props) {
     if (date?.isValid) {
       // Compute the moment the token expired: the configured number of days after the visit, at midnight
-      date = date.plus({days: config?.allowedPostVisitCompletionTime || 0}).endOf('day');
+      date = date.plus({days: tokenLifetime || config?.allowedPostVisitCompletionTime || 0}).endOf('day');
```
If the value of tokenLifetime is 0 it will be evaluated to false and end up being ignored. Use the ?? operator instead of ||.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/ValidateCredentialsServlet.java:
```diff
@@ -407,7 +408,16 @@ private boolean isVisitUpcoming(final Session session, Node visitInformationForm
+                        this.formUtils,
+                        visitSubject,
+                        "/Questionnaires/Visit information/surveys",
+                        "tokenLifetime",
```
It seems unnecessary to specify the property name here, it should be set internally to the getTikenLifetime method.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/surveytracker/SurveyTracker.java:
```diff
@@ -182,15 +183,16 @@ private void handleResourceEvent(final ResourceChange event)
                 updateSurveySubmittedDate(node, session);
             } else if (isAnswerForVisitTime(node)) {
-                updateSurveyExpirationDate(node, session);
+                updateSurveyExpirationDate(node, node, session);
```
Why is the same node passed twice?

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Minor code optimisation

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Better logical operator use to avoid mistreating 0 values

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Fixed typo in node usage

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Added tokenLifetime to clinic UI

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Avoid defaulting tokenLifetime to 0 in UI;
Save tokenLifetime via UI;

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Fixed clinic creation with empty token

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Fixed clinic creation with non-empty token

veronikaslc commented:
Tested listed scenarious, should be good to go for @sdumitriu pre-review

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/AbstractEmailNotification.java:
```diff
@@ -118,10 +117,11 @@ public long sendNotification(final int differenceInDays, final EmailTemplate tem
                 Node appointmentForm = this.formUtils.getForm(appointmentDate);
-                if (appointmentForm == null || isSurveyCompleted(appointmentForm, session)) {
+                Node visitSubject = this.formUtils.getSubject(appointmentForm, "/SubjectTypes/Patient/Visit");
+                if (appointmentForm == null || AppointmentUtils.getVisitSurveysComplete(this.formUtils, visitSubject)) {
```
`geetVisitSurveysComplete` should be renamed to `isVisitSurveyComplete` to comply with the naming convention for methods returning a boolean value.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/AbstractEmailNotification.java:
```diff
@@ -160,12 +160,17 @@ private Email renderTemplate(final EmailTemplate template, final Node appointmen
-        tokenExpiryDate.add(
-            Calendar.DATE,
-            this.patientAccessConfiguration.getAllowedPostVisitCompletionTime());
+        final int postVisitCompletionTime = this.patientAccessConfiguration.getAllowedPostVisitCompletionTime();
```
Let's call this `defaultTokenLifetime` to make the code easier to read.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/AbstractEmailNotification.java:
```diff
@@ -160,12 +160,17 @@ private Email renderTemplate(final EmailTemplate template, final Node appointmen
+        final int tokenLifetime = AppointmentUtils.getTokenLifetime(
+            this.formUtils,
+            visitSubject,
+            "/Questionnaires/Visit information/clinic",
```
We don't need to have this as a parameter. When obtaining the token lifetime for a visit, we will always look at the configuration from the visit's clinic. let's make it internal to the `AppointmentUtils#getTokenLifetime` method.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/ValidateCredentialsServlet.java:
```diff
@@ -407,7 +408,15 @@ private boolean isVisitUpcoming(final Session session, Node visitInformationForm
                 try {
                     Calendar limit = (Calendar) p.getDate().clone();
-                    limit.add(Calendar.DATE, this.patientAccessConfiguration.getAllowedPostVisitCompletionTime());
+                    final int postVisitCompletionTime =
+                        this.patientAccessConfiguration.getAllowedPostVisitCompletionTime();
+                    Node visitSubject = this.formUtils.getSubject(visitInformationForm, "/SubjectTypes/Patient/Visit");
+                    final int tokenLifetime = AppointmentUtils.getTokenLifetime(
+                        this.formUtils,
+                        visitSubject,
+                        "/Questionnaires/Visit information/clinic",
+                        postVisitCompletionTime);
+                    limit.add(Calendar.DATE, tokenLifetime);
```
Same comments regarding naming and parameters as in AbstractEmailNotification utils.

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Better naming for params

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/AppointmentUtils.java:
```diff
@@ -247,14 +228,40 @@ public static String getValidClinicEmail(FormUtils formUtils, Node formRelatedSu
+     * @param defaultLifetime the default to return
+     * @return the token lifetime in days
+     */
+    public static int getTokenLifetime(FormUtils formUtils, Node formRelatedSubject,
```
I don't like that this method is in `patients/emailnotifications`.  Instead, I would replace `PatientAccessConfiguration.getAllowedPostVisitCompletionTime()` with two methods, `getDefaultAllowedPostVisitCompletionTime()` and `getAllowedPostVisitCompletionTime(Node visitInformationNode)`.

marta- replied:
I don't think we need to rename to getDefault..., a simple overload would be more elegant in my opinion

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Code refactoring

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/Clinics.jsx:
```diff
@@ -43,7 +43,8 @@ function Clinics(props) {
     "survey": "Surveys",
-    "emergencyContact": "Emergency Contact"
+    "emergencyContact": "Emergency Contact",
+    "tokenLifetime": "Token Lifetime"
```
This is too technical, it should be similar to how it's displayed in the Patient Access admin page, including a description.

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/Clinics.json:
```diff
@@ -5,5 +5,6 @@
   "survey": "string",
   "emergencyContact": "string",
   "description": "string",
+  "tokenLifetime": "double",
```
Too technical, just like the PatientAccess setting it overrides, it should be named `allowedPostVisitCompletionTime`.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/api/PatientAccessConfiguration.java:
```diff
@@ -48,7 +50,17 @@ public interface PatientAccessConfiguration
+     * Returns the token lifetime associated with the clinic linked to the Subject
+     * related to the visitInformationNode Resource or default if it cannot be found.
+     *
+     * @param visitInformationNode the JCR visit Information Node
```
```suggestion
     * @param visitInformationNode the JCR Visit Information Node
```

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/AbstractEmailNotification.java:
```diff
@@ -118,15 +117,17 @@ public long sendNotification(final int differenceInDays, final EmailTemplate tem
                 Node appointmentForm = this.formUtils.getForm(appointmentDate);
-                if (appointmentForm == null || isSurveyCompleted(appointmentForm, session)) {
+                Node visitSubject = this.formUtils.getSubject(appointmentForm, "/SubjectTypes/Patient/Visit");
+                if (appointmentForm == null || AppointmentUtils.isVisitSurveyComplete(this.formUtils, visitSubject)) {
```
This will have to be changed later, since we will have different visit information forms for the same subject.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/api/PatientAccessConfiguration.java:
```diff
@@ -48,7 +50,17 @@ public interface PatientAccessConfiguration
      * @return A number of days
      */
-    int getAllowedPostVisitCompletionTime();
+    int getDefaultAllowedPostVisitCompletionTime();
```
I don't think it's necessary to rename this function, just overload it.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/api/PatientAccessConfiguration.java:
```diff
@@ -48,7 +50,17 @@ public interface PatientAccessConfiguration
+     *
+     * @return A number of days
+     */
+    int getAllowedPostVisitCompletionTime(Node visitInformationNode);
```
I would rename `visitInformationNode` to `visitInformationForm` to make it more clear what kind of node we are passing

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientAccessConfigurationImpl.java:
```diff
@@ -121,17 +129,36 @@ public boolean isPatientIdentificationRequired()
+        {
+            Node visitSubject = this.formUtils.getSubject(visitInformationNode, "/SubjectTypes/Patient/Visit");
+            Node clinicNode = AppointmentUtils.getValidClinicNode(this.formUtils, visitSubject,
+                AppointmentUtils.CLINIC_PATH);
```
CLINIC_PATH should not be a parameter. There's only one way to obtain the visit clinic, and that is by looking at the `/Questionnaires/Visit information/clinic` question.

veronikaslc replied:
Then why do we parametrise it [here](https://github.com/data-team-uhn/cards/blame/dev/modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/AppointmentUtils.java#L195) and [here](https://github.com/data-team-uhn/cards/blob/dev/modules/patient-portal/src/main/java/io/uhndata/cards/patients/emailnotifications/EmailFormAlerts.java#L96)?

marta- replied:
Whatever the reason was at the time when that code was written, it appears to be obsolete now. Maybe @sdumitriu and @IntegralProgrammer can help figure it out?

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Code refactoring

sashaandjic commented:
After initial Visit info creation I got the following token expirations:
- Clinic A (blank) --> 29 days as expected
- Clinic B (2) --> 4 days ( 1 day expected)
- Clinic C (-2) --> 0 days ( I was expecting yesterday)
- 
![image](https://github.com/data-team-uhn/cards/assets/23243244/e0cd4dff-823e-4708-b925-a5a777a569fe)

sashaandjic commented:
After initial visit creation I got the following expiry dates:
- Clinic A ( token lifetime blank) --> 29 days as expected
- Clinic B ( token lifetime 2) --> 4 days ( I expected 1, tomorrow)
- Clinic C ( token lifetime -2) -> 0 days ( I expected yesterday)
![image](https://github.com/data-team-uhn/cards/assets/23243244/64177e78-bca2-4726-b440-a52ebc03daf2)

sashaandjic commented:
After manually generating tokens I got the same dates, however for Clinic A the start page is stating 30 days instead of 29:
![image](https://github.com/data-team-uhn/cards/assets/23243244/b27b402e-ddd8-4796-9673-a332cd3e051b)

![image](https://github.com/data-team-uhn/cards/assets/23243244/e78f9d9b-b1ae-47aa-be43-33fa2747c0f0)

![image](https://github.com/data-team-uhn/cards/assets/23243244/39c8acf5-e51e-412f-b4d4-301c84930147)

sashaandjic commented:
Changing token life in composum to 0, renders "Surveys expires in 7 hours".
![image](https://github.com/data-team-uhn/cards/assets/23243244/594d62de-8c73-40e1-9dd7-c21f6002f081)
![image](https://github.com/data-team-uhn/cards/assets/23243244/7d72a120-8ddb-4dbe-8872-cf4bebc3c47a)

sashaandjic commented:
here is what happens in dev. expiry day is 32 days from the discharge date. for juy 1st discharge, Aug 2nd is the token expiry date. There are 5 days until Aug 2nd, but the start screen is saying 3. So the problem is in the dev branch.
![image](https://github.com/data-team-uhn/cards/assets/23243244/b5028bc3-f628-4f85-b037-4fc862122a21)

sashaandjic commented:
This is interesting: the form is showing Aug2 and in 
 http://localhost:8080/jcr:system/cards:tokens.30.jhttp://localhost:8080/jcr:system/cards:tokens.30.son` we stored Aug 1
![image](https://github.com/data-team-uhn/cards/assets/23243244/e9a97cf8-cf5d-42f3-b67d-7a952b49bd05)

sdumitriu commented:
> @sashaandjic The code that adds days is from dev. @sdumitriu is there a logic behind[ those lines](https://github.com/data-team-uhn/cards/blame/dev/modules/patient-portal/src/main/java/io/uhndata/cards/patients/surveytracker/SurveyTracker.java#L237C17-L238C54) that add 2 more days?

I think adding one more day on line 238 is a mistake, but adding one day on line 237 is kind of OK, unless we change the definition of what the "post visit completion time" setting does.

If the visit is today at 9:00 AM, and completion time = 0, then we actually want to allow the patient to submit until the end of the day, so exactly midnight _tomorrow_. That is why we add one more day, and reset all time fields to 0.

sdumitriu commented:
And I'm not sure but I think that there may be one more day added/removed in the UI depending on what kind of rounding the "approximate time" library uses, e.g. if it's 09:00 now, is until 00:00 two days from now (or 24:00 tomorrow) one, two, or three days? I think it always rounds down in whole days, so it's just one day away.

But how do normal people think if they see "the link expires in 1 day"? I would interpret it as "I have at least 24 more hours", so somewhere beyond 9:00 tomorrow, probably at midnight.

veronikaslc commented:
@sdumitriu I agree with you on the latter one. As a user I'd think I have 1 full day.

veronikaslc commented:
@sdumitriu Adding one more day in `SurveyTracker.java` still creates the discrepancy between token expiry days count on welcome screen and survey expiry date in Survey event form.

sdumitriu reviewed (commented):
I just noticed I had a pending review, not sure if these comments still make sense.

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Addressed codereview comments

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Updated the wording

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Adressed the checkstyle violations

sashaandjic reviewed (approved):
As per instructions above:
![image](https://github.com/data-team-uhn/cards/assets/23243244/dced8404-46d5-401b-aaf7-a473b37fb327)
-----
Once I changed in  /bin/browser token lifetime of clinic B to be 0 I got "This survey will expire in 5 minutes":
![image](https://github.com/data-team-uhn/cards/assets/23243244/69d9c462-90db-4e62-817f-c82a32b1f7de)
-------
... but when I tried the same 20 minutes later, I got the same?!?
![image](https://github.com/data-team-uhn/cards/assets/23243244/cd99265f-b8fa-4b27-b4fe-c7fa8edd6cc0)
-----

sashaandjic commented:
I am perplexed as why all 3 tokens have the same expiry date in http://localhost:2286/jcr:system/cards:tokens.20.json
![image](https://github.com/data-team-uhn/cards/assets/23243244/516b77fe-0937-4c7c-a02c-706df4fcd9a3)

marta- commented:
> I am perplexed as why all 3 tokens have the same expiry date ( in 29 days) in 
> 
> I played more with this and generated bunch of new tokens. As a patient I get the expected behavior ( for clinic A 29 days, and for clinic B in 5 minutes), but in json all my tokens are going to expire in 29 days. The other thing that I noticed is that in json we store part of the token in `jcr:uuid`.
> 

Looks like the expiry date does not take into account the clinic setting when the token is generated via `token.html`.  @veronikaslc please update here: https://github.com/data-team-uhn/cards/blob/5da9b83e1166f2c2b3ca990a9b3fdc15bc06a389/modules/token-authentication/src/main/resources/SLING-INF/content/libs/cards/Subject/token.html.esp#L32

Also, the branch needs a rebase and conflict resolution.

marta- commented:
> ## Once I changed in /bin/browser token lifetime of clinic B to be 0 I got "This survey will expire in 5 minutes":
> 
> ## ... but when I tried the same 20 minutes later, I got the same?!?
> 

I don't see any expiration date in the posted screenshots, The only "5 minutes" I see is the estimate for how long it will take to fill out the survey.

No expiry info on the screen most likely means these lines were executed in both cases, which doesn't seem right:

https://github.com/data-team-uhn/cards/blob/dev/modules/patient-portal/src/main/frontend/src/patient-portal/QuestionnaireSet.jsx#L599-L601

.

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Take into account the clinic setting when the token is generated via token.html

marta- commented:
There's a mismatch between the label used in the clinics table and the one used in the creation dialog. One of them could use a refactoring.

![2307](https://github.com/data-team-uhn/cards/assets/651980/b731104b-6d88-4a5c-b7c9-025ead50188c)

marta- commented:
> There's a mismatch between the label used in the clinics table and the one used in the creation dialog:
> 
> The whole "Clinics" code could use a refactoring, as this new mismatch is not singular.

For now, to work around the lack of support for custom labels in the dialog, I would suggest the following quick fixes:
* get rid of [`formattedNames`](https://github.com/data-team-uhn/cards/pull/1562/files#diff-138230640769144a52b96d11869fd6db2c8c307f55e5739fbff97173e6938e00R41-R48) from Clinics.jsx lines 41-48 and use [`camelCaseToWords`](https://github.com/data-team-uhn/cards/blob/9546f738e0f6a65661928f9851f3393e86362a39/modules/data-entry/src/main/frontend/src/questionnaireEditor/LabeledField.jsx#L34-L36) to generate the labels, for consistency between the Clinics table and the onboarding dialog;
* rename (again) `allowedPostVisitCompletionTime` to `nbOfDaysRelativeToEventToCompleteSurvey`
* optional, if the label is still not clear enough, add a `Clinics-hints.json` with a dint for this field _("Use a negative number when patient responses are due a number of days before the event, 0 for the day of the event, and a positive number when their responses are expected after the event.")_ and pass it to  the `Fields` component within the dialog.

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Some work around the lack of support for custom labels in the New clinic mapping dialog

marta- commented:
> * optional, if the label is still not clear enough, add a `Clinics-hints.json` with a dint for this field _("Use a negative number when patient responses are due a number of days before the event, 0 for the day of the event, and a positive number when their responses are expected after the event.")_ and pass it to  the `Fields` component within the dialog.

The Clinics admin screen makes more sense now in my opinion, but I think we do need the hints for the dialog, let's add them please @veronikaslc

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Added hints file for the nbOfDaysRelativeToEventToCompleteSurvey field

veronikaslc commented:
Added hints and rebased on the latest `dev`.

marta- commented:
> Added hints and rebased on the latest `dev`.

@veronikaslc the hints json file is missing from the last commit.

marta- reviewed (changes requested):
Once again, variable naming is the biggest challenge... @sdumitriu requested another rename, and we settled on `daysRelativeToEventWhileSurveyIsValid`. Please also rename the getter `getAllowedPostVisitCompletionTime` from PatientAccessConfiguration.java and PatientaccessConfigutaionImpl.java to match the new field name @veronikaslc

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Better naming

sashaandjic commented:
Unable to generate token:
![image](https://github.com/data-team-uhn/cards/assets/23243244/45d71d35-0633-4988-96c9-0eb4cfe286b4)

marta- commented:
> Unable to generate token:

Looks like [this simple fix](https://github.com/data-team-uhn/cards/pull/1562/commits/dbfc36b8a0a1cc653cb14d794968e6240d519a6c) is not working as expected...

sashaandjic commented:
Oh, hold on, I did not have the latest version
![image](https://github.com/data-team-uhn/cards/assets/23243244/0111790c-80d2-4eab-afe5-6f4bea223321)

sashaandjic commented:
Bug is still there
`500
TypeError: Cannot call property getParent in object Sun Aug 27 2023 19:25:00 GMT-0400. It is not a function, it is`

veronikaslc committed:
CARDS-2286: Allow to configure token life time for each survey
Bugfix in getting visit information form for token generation

sashaandjic reviewed (approved):
tested
As per steps described in the testing instructions:

Clinic A
-----
![image](https://github.com/data-team-uhn/cards/assets/23243244/aa7ab886-2998-45c5-be11-183b61b1be84)

Clinic B
-----
![image](https://github.com/data-team-uhn/cards/assets/23243244/5103527b-c24c-4eb6-80de-608be07c3ee8)

After changing number of days to response to 0 for Clinic B:
-----
![image](https://github.com/data-team-uhn/cards/assets/23243244/8996e4f9-c354-45c2-bb37-21f7b4e49db8)

Token expiry date in json:
------
![image](https://github.com/data-team-uhn/cards/assets/23243244/9a68ff6f-9a89-409b-8216-a55716955763)

Clinic dialog:
-----
![image](https://github.com/data-team-uhn/cards/assets/23243244/40836fb5-a46e-4387-ab59-270af89b04da)

Clinic list:
-----
![image](https://github.com/data-team-uhn/cards/assets/23243244/c5ad23c3-3f67-47c7-9948-56f1e714912e)

sdumitriu reviewed (approved):
Still testing, but code looks good.

sdumitriu committed:
CARDS-2286: Allow to configure token life time for each survey
Bugfix: unsubmitted forms are never deleted
